### PR TITLE
[jit][script] Allow tuples to be re-assigned

### PR DESCRIPTION
**Pull request by zdevito**
(Just read the last commit, this is layered on top of my other PR).

This turned out to be necessary for FairSeq to work correctly.

This commit improves our support of tuples by making them more first-class.
In particular, it allows tuples to be re-assigned across loops and ifs.
It does this by making them first-class values in the Graph IR, and then
removing the tuples in a LowerTuples pass.

An alternative approach would have added more support for desugaring tuples
in the Environment object as they were emitted. Instead,
the current approach was chosen anticipating a future when tuples are
fully supported (including the interpreter). In that future, the current
code can be completly reused with the LowerTuples pass just becoming
a optimization that removes unneeded tuple allocations.

### CODE REVIEW

**Comment by zdevito** [comment]
@jamesr66a @apaszke

**Review by apaszke — APPROVED** [review]

File: test/test_jit.py [hunk 1]
```diff
@@ -2383,6 +2393,63 @@ def forward(self, x, seq_lens):
...
+            t0, t1 = c
+            while False:
+                t0, t1 = c
+                c = torch.chunk(i, dim=0, chunks=2)
+            return t0
+
+        v = torch.rand(10, 3)
+        self.assertEqual(torch.chunk(v, dim=0, chunks=2)[0], foo(v))
+
+        with self.assertRaisesRegex(RuntimeError, "variable 'a' previously has type"):
```
Comment (jamesr66a) [inline test/test_jit.py]: We're diverging from Python semantics here, I think. Is this alright?

Comment (apaszke) [reply test/test_jit.py]: It's ok. We're only a subset, and it's impossible to figure out a type of this variable if we don't know a priori whether the `if` branch will run.

File: torch/csrc/jit/passes/dead_code_elimination.cpp [hunk 2]
```diff
@@ -6,7 +6,7 @@ void EliminateDeadCode(std::shared_ptr<Graph>& graph) {
   EliminateDeadCode(graph->block());
 }
 bool hasSideEffects(Node * node) {
-  return node->kind() == prim::Print;
+  return node->kind() == prim::Print || node->blocks().size() > 0;
```
Comment (jamesr66a) [inline torch/csrc/jit/passes/dead_code_elimination.cpp]: Why is this needed? Because we can reference values implicitly from blocks?

File: torch/csrc/jit/passes/lower_tuples.cpp [hunk 3]
```diff
@@ -0,0 +1,114 @@
...
+  prim::If,
+  prim::Loop,
+  prim::TupleUnpack,
+  prim::TupleConstruct,
+  prim::Param,
+  prim::Return,
+ };
+
+
+static void LowerTuples(Block* block);
```
Comment (jamesr66a) [inline torch/csrc/jit/passes/lower_tuples.cpp]: Why are these functions static?

File: torch/csrc/jit/passes/dead_code_elimination.cpp (hunk 2 above)
Comment (apaszke) [reply torch/csrc/jit/passes/dead_code_elimination.cpp]: I think previously we could have removed an `If` node that has no outputs (but its body can have side effects!).

File: torch/csrc/jit/passes/lower_tuples.cpp (hunk 3 above)
Comment (apaszke) [reply torch/csrc/jit/passes/lower_tuples.cpp]: Aren't they internal to the pass? Only `void LowerTuples(Graph& graph)` should be public

Comment (jamesr66a) [reply torch/csrc/jit/passes/lower_tuples.cpp]: Is there a reason to use `static` over an anonymous namespace?

**Comment by jamesr66a** [comment]
@zdevito are you going to add support for the NameConstant node in the frontend?

**Comment by apaszke** [comment]
@jamesr66a What's a NameConstant?

**Comment by jamesr66a** [comment]
@apaszke this: http://greentreesnakes.readthedocs.io/en/latest/nodes.html#NameConstant

`True`, `False`, and `None` appear as NameConstant nodes in 3.4+. This is what is causing the CI failure (and also failing locally for me since I use py3)

File: torch/csrc/jit/script/compiler.cpp [hunk 4]
```diff
@@ -166,7 +173,9 @@ struct Environment {
 
     if (!retval && (retval = findInParentFrame(ident)) &&
         getBlockOwningKind() == prim::Loop) {
-      retval = createCapturedInput(ident);
+      if(Value* simple_val = asSimple(retval)) {
+        retval = createCapturedInput(simple_val, ident);
+      }
```
[the remark this answers is not in the document]
Comment (zdevito) [reply torch/csrc/jit/script/compiler.cpp]: Tuple values are first-class in the graph, so they are SimpleValues and are captured. This was a bug in the last commit: we shouldn't try to capture other SugaredValues. The reason we create a captured input in getVar is because we are concerned that a future setVar will cause it to be come a loop-carried dependency. For SugaredValues, this case is prevented in setVar (you cannot reassign a sugared value).

File: torch/csrc/jit/script/compiler.cpp [hunk 5]
```diff
@@ -490,16 +506,13 @@ struct to_ir {
 
     // Register outputs in each block
     for (const auto& x : sorted_mutations) {
-      true_block->registerOutput(save_true->getVar(x, stmt.range()));
-    }
-    for (const auto& x : sorted_mutations) {
-      false_block->registerOutput(save_false->getVar(x, stmt.range()));
+      auto tv = save_true->getVar(x, stmt.range());
+      true_block->registerOutput(tv);
+      auto tf = save_false->getVar(x, stmt.range());
```
[the remark this answers is not in the document]
Comment (zdevito) [reply torch/csrc/jit/script/compiler.cpp]: lol, yes....
I mean, maybe it stands for `totally verified` and `totally false`....

File: test/test_jit.py (hunk 1 above)
Comment (zdevito) [reply test/test_jit.py]: Yeah, thankfully this fully fits into "a subset of programs, which don't change the type of their variables, will work"

File: torch/csrc/jit/passes/dead_code_elimination.cpp (hunk 2 above)
Comment (zdevito) [reply torch/csrc/jit/passes/dead_code_elimination.cpp]: Correct. I hit this while debugging and saw my print inside and if disappear. We really should change to the version @ezyang suggested where side-effecting operators like print take and return a "world" token. which would then become an output of the if-statement and then we can remove the `hasSideEffects` check.

**Comment by zdevito** [comment]
NamedConstant issues are fixed in the other PR, and I will rebase this one.

File: torch/csrc/jit/script/compiler.cpp (hunk 5 above)
[the remark this answers is not in the document]
Comment (apaszke) [reply torch/csrc/jit/script/compiler.cpp]: That totally makes sense

File: torch/csrc/jit/passes/lower_tuples.cpp (hunk 3 above)
Comment (apaszke) [reply torch/csrc/jit/passes/lower_tuples.cpp]: Not really, a matter of taste I guess

**Review by apaszke — DISMISSED** [review]
Code LGTM, but you accidentally changed the ONNX subtree. Feel free to dismiss this review once this is fixed

**Review by gchanan — APPROVED** [review]
Assuming this good given @apaszke review
